CARDS-1006: When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts

veronikaslc opened the pull request.

Commit:
CARDS-1006: When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts

sashaandjic commented:
Problem found. Steps to reproduce:
- create a new Patient Info form
- add DOB then  click on `Save & view`
- all offending fields are listed as expected
![image](https://user-images.githubusercontent.com/23243244/136265950-3f6e0251-1dfb-464f-8aad-867fa95f2ba8.png)
- edit and select gender then `Save & view'
Gender is still marked as an offending field
![image](https://user-images.githubusercontent.com/23243244/136266116-765943e7-703f-4e45-8033-a8cb97ea9529.png)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -397,6 +429,15 @@ function Form (props) {
+                 className={`${classes.subjectChip} ${classes[item.statusFlags[0] + "Chip"] || classes.DefaultChip}`}
+                 size="small"
+               />
+          })}
```
There must have been a misunderstanding / unclear specs. Here we should display the form's `statusFlag` (which can currently be an empty array or contain one or both of "Invalid","Incomplete"), just like we do on the [Subject chart](https://github.com/data-team-uhn/cards/blob/dev/modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx#L516). We should not list all the invalid/missing fields.

Example of expected display:
![image](https://user-images.githubusercontent.com/651980/136286624-7c60d757-e401-4440-8801-44b62792f85f.png)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -271,9 +273,17 @@ function Form (props) {
   let onClose = (event) => {
     // Redirect the user to the view form mode
     // ...but only after the Form has been saved and checked-in
+    let hash = window.location.hash ? indow.location.hash.substr(1) : '';
```
Typo: 
```suggestion
    let hash = window.location.hash ? window.location.hash.substr(1) : '';
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -141,7 +129,7 @@ function Section(props) {
   if (isEdit) {
     collapseClasses.push(classes[displayMode + 'Section']);
   }
-  if (isEdit && !displayed || !isEdit && !hasAnswers) {
```
Now sections that don't have any incomplete questions are displayed empty in view mode.

To reproduce:
* create a new questionnaire Q
* create a section and add a title to it - "The section"
* create a question in that section (variable name q, text "The question") and leave everything else as selected by default; this question cannot have an invalid/incomplete answer
* create a form from the new questionnaire, do not answer The question, click save and go to view mode; result => "The section" appears displayed in view mode.

The issue can also be noticed in the "Patient information" form in view mode. There is a larger gap between "Mutation-Function" and "Age at last follow up". The gap corresponds to the empty testing section that should not have been rendered.

` || !isEdit && !hasAnswers` should be expanded with checking the `statusFlag` of the answer section instead of being completely removed. If we're in view mode and do not have any answers and the section is not marked as incomplete, then hide it.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -271,9 +273,17 @@ function Form (props) {
     // Redirect the user to the view form mode
     // ...but only after the Form has been saved and checked-in
+    let hash = window.location.hash ? indow.location.hash.substr(1) : '';
+    // Upon opening the invalid/incomplete form in view mode, scroll to the first invalid/incomplete question
```
The wrong mode was specified in the task description. It should be "Upon opening the invalid/incomplete form in EDIT mode, scroll to the first invalid/incomplete question". The reasoning behind it is so that the user can go straight to fixing the wrong/unanswered question. The specs are now updated, please update the code as well.

The hash should be discarded when going back to view mode.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -44,7 +44,7 @@ function Question (props) {
       <CardContent className={isEdit ? classes.editModeAnswers : classes.viewModeAnswers}>
         <div className={compact ? classes.compactLayout : null}>
-        { isEdit && !disableInstructions &&
+        { (!isEdit || !disableInstructions) &&
```
This is part of what is causing the bug reported by @sashaandjic . 

`disableInstructions` is here for questions where the instructions coloring based on existing answers are handled directly in the question. IIRC this is the case for questions that use the MultipleChoice component and Number range questions. If we display the instructions here for those questions, there's not enough information to establish if the answer is acceptable, so the instructions will always be colored in red.

What I would suggest is:
* Instead of checking for `isEdit` here (or in any other component displaying instructions), pass it on to the `AnswerInstructions` component
* In the `AnswerInstructions` component, render the instructions if `isEdit` or [`answerIsAcceptable` ](https://github.com/data-team-uhn/cards/blob/dev/modules/data-entry/src/main/frontend/src/questionnaire/AnswerInstructions.jsx#L32)is `false`(the latter is for view mode where the answer is incomplete)
*  We may need to tweak MultipleChoice and NumberQuestion (and any other component rendering AnswerInstructions itself) to ensure enough information about the existing answer passed on to AnswerInstructions.

Commit:
CARDS-1006: When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts
Fixed section show in view mode and focus on edit mode

Commit:
CARDS-1006: When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts
Fixed answer instructions showed in view mode for answered questions

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/AnswerInstructions.jsx:
```diff
@@ -49,7 +50,7 @@ function AnswerInstructions (props) {
   }, [currentAnswers]);
 
-  return (instructionsExist && (
+  return (instructionsExist && !answerComplete && (
```
Bug: All `AnswerInstructions` have now disappeared form edit mode.

marta- replied:
Addressed in 84f33e7 .

marta- committed:
CARDS-1006: When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts
* Fixed regression introduced in d5b71df: answer instructions no longer present in edit mode
* Minor refactoring for better code readability

sashaandjic commented:
Problem of a cosmetic nature found. Steps to reproduce:
- create a new Radio form
- set `Radiation therapy =Yes`
- save & view --> all good
![image](https://user-images.githubusercontent.com/23243244/137042933-93d1fbba-1a46-4dfc-b544-ca94a51ed428.png)
- edit
- orange box is displayed around the next required field `Radiation field` --> **correct** 
- add value to the `Radiation field`
![image](https://user-images.githubusercontent.com/23243244/137043003-f233344c-0486-4d17-8c43-39f30c884d05.png)
- save & view -> the orange box is again around `Radiation field` - **wrong**
![image](https://user-images.githubusercontent.com/23243244/137043088-c024828f-c75b-4ffa-b8ca-2e4098e9620e.png)
- edit --> oranhe box didn't move to the next required field `Radiation source` - **wrong**  
![image](https://user-images.githubusercontent.com/23243244/137043266-e8397f85-31f6-4972-b7ee-b507780ed215.png)

The orange box seems to be stuck on the `Radiation field` even after the form status is complete. Refresh does not remove the orange box. The only way to get rid of it is by moving away from the form and opening it again.

marta- committed:
CARDS-1006: When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts
Minor refactoring for better code readability

marta- committed:
[misc] Better styling for highlighted items in forms and questionnaires

marta- committed:
CARDS-1006: When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts
* Bugfix: once an incomplete question is highlighted for editing, the highlight persists until refreshing the page even if a valid answer is entered
* Improvement: use "error" color for highlighting invalid/incomplete questions for editing
* Side effect: when an invalid question is focused/highlighted for editing, changing anything in the form will remove the highlight by removing the location hash, and consequently and trigger a save (because of the location change)

sashaandjic commented:
Tested the lattes changes -all good

Waiting for the final changes that we discussed on the stand-up today: clickable `Incomplete `chip will take user to the first blank required field + possibility to take user to the next one after filling the current required field with proper value
Now I am thinking we should do the same with `Invalid` forms. Separate task?

marta- committed:
CARDS-1006: When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts
Removed the highlighting of the first missing field when clicking edit, since the current implementation messes up form versioning

marta- commented:
The last commit removes the highlighting of the first missing field when clicking edit, since with the current implementation of field highlighting that can mess up form versioning.

sashaandjic reviewed (approved):
tested

sashaandjic commented:
QUESTION: the boxes appear much higher in view mode. 
Can we get rid of that extra space? 
The header shrinks when `Incomplete` chip is not there. Can we do the same when `This question is mandatory` isn't there?

![image](https://user-images.githubusercontent.com/23243244/144107221-077a4361-60fe-40b0-9231-b596e993bd4b.png)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -52,7 +52,6 @@ const questionnaireStyle = theme => ({
     informationCard: {
     },
     viewModeAnswers :{
-      paddingTop: theme.spacing(0),
```
> QUESTION: the boxes appear much higher in view mode.
> Can we get rid of that extra space?
> The header shrinks when Incomplete chip is not there. Can we do the same when This question is mandatory isn't there?

@veronikaslc Do you remember why you removed this line?

marta- replied:
It appears to be so that the instructions we display in view mode (e.g. This question is mandatory) don't overlap the question text. This can be achieved by adjusting the margin of the instructions in view mode instead.

sdumitriu replied:
Is this going to be addressed in this PR?

sashaandjic replied:
No. I will make a separate task.

sashaandjic replied:
https://phenotips.atlassian.net/browse/CARDS-1536

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/AnswerInstructions.jsx:
```diff
@@ -29,9 +29,10 @@ import QuestionnaireStyle from "./QuestionnaireStyle";
+  let [ answerIsAcceptable, setAnswerAcceptable] = useState();
 
-  const instructionsExist = minAnswers > 0 || maxAnswers > 1;
+  const instructionsExist = (minAnswers > 0 || maxAnswers > 1) && (isEdit || existingAnswer?.[1]?.statusFlags?.length > 0);
```
This assumes that all status flags are bad, which is not a valid assumption. For example, manually marking an answer with a `validated` flag will cause it to display the answer instructions for it in view mode, even if nothing is wrong with the answer.

On one hand, we should not assume that all status flags are bad and indicate a problem with how the question was answered.

On the other hand, we should not hard-code the list of flags that are bad, since we may add other flags besides "incomplete" and "invalid" that will mark a problem with the answer.

I guess given the flags that we do have right now we can merge this as is, and revisit once we actually introduce new flags.

marta- replied:
This assumption is made in quite a few places. +1 to leave as is for this PR and rework later.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/AnswerInstructions.jsx:
```diff
@@ -46,7 +47,7 @@ function AnswerInstructions (props) {
 
   useEffect(() => {
-    setAnswerAcceptable((currentAnswers >= minAnswers) && (!(maxAnswers >= minAnswers) || currentAnswers <= maxAnswers))
+    setAnswerAcceptable((currentAnswers >= minAnswers) && (!(maxAnswers >= minAnswers) || currentAnswers <= maxAnswers) || !isEdit && existingAnswer?.[1]?.statusFlags?.length == 0)
```
Same here.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -300,6 +302,28 @@ function Form (props) {
 
+  useEffect(() => {
+    if (data) {
+      let incompletes = getIncompleteAnswers(data, []);
+      setIncompleteAnswers(incompletes);
+    }
+  }, [data])
+
+  let getIncompleteAnswers = (obj, data) => {
```
I think naming two different things `data`, especially when the outer `data` is **not** passed as the `data` argument, is prone to induce headscratching. A better name for the `data` function parameter would be `incompletes`.

marta- replied:
Renamed second argument in 9e7e9f2.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -31,10 +31,12 @@ function Question (props) {
   let { classes, children, questionDefinition, existingAnswer, isEdit, preventDefaultView, defaultDisplayFormatter } = props;
   let { text, compact, description, disableInstructions } = { ...questionDefinition, ...props }
 
+  let [ hasInvalidAnswer, setHasInvalidAnswer ] = React.useState(existingAnswer?.[1].statusFlags?.length > 0);
```
Again, not all flags mean an answer is invalid.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -31,10 +31,12 @@ function Question (props) {
     <Card
       variant="outlined"
-      className={classes.questionCard}
+      className={classes.questionCard + (hasInvalidAnswer ? (' cards-with-invalid-answer') : '')}
```
Shouldn't the second part also be taken from `classes` instead of a raw classname?

marta- replied:
This is actually a leftover, removed in commit 382d722.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx:
```diff
@@ -114,7 +114,8 @@ function Section(props) {
 
-  const isDisplayed = isEdit && displayed || !isEdit && hasAnswers;
+  // Display the section in view mode if it has answers or is marked as incomplete
+  const isDisplayed = isEdit && displayed || !isEdit && (hasAnswers || existingAnswer[0]?.[1].statusFlags?.length > 0);
```
Why are there different `isDisplayed` and `displayed` properties?

I think both of them have misleading names, since I would expect this to mean that the section is actually displayed in the page right now, but even if both are true, `pageActive === false` can still hide the section.

`displayed` means "the condition for displaying the section is satisfied", so I would name it `conditionSatisfied`.

`isDisplayed` means "the section should be included in the output, either displayed or as a valid page", so I would name it `displayable`.

But this rename is for a different PR.

marta- replied:
Some renaming is done in #823 .

marta- committed:
CARDS-1006: When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts
Removed error highlighting that would never be triggered after the changes from the previous commit

marta- committed:
CARDS-1006: When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts
Renamed parameter for better code readability

marta- commented:
Rebased on top of the latest dev and addressed some of @sdumitriu 's refactoring comments. No behavior should have changed. @sashaandjic please retest.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormEntry.jsx:
```diff
@@ -57,8 +59,21 @@ export const ENTRY_TYPES = QUESTION_TYPES.concat(SECTION_TYPES).concat(INFO_TYPE
+
+  // if autofocus is needed and specified in the url
+  useEffect(() => {
+    setAnchor(decodeURIComponent(location.hash.substr(1)))
```
Replace deprecated function call:
```suggestion
    setAnchor(decodeURIComponent(location.hash.substring(1)))
```

marta- committed:
[misc] Replace deprecated function

sashaandjic reviewed (approved):
tested
